### DISCUSSION TITLE
3.12 compatibility issues with `fuse_loop_config`

### DISCUSSION THREAD

**Discussion by overheadhunter** [libfuse/libfuse#701] [Q&A]
A filesystem compiled against an older 3.x API will fail at runtime when invoking `fuse_loop_mt()`, if the system ships with `libfuse3.so.3.12.0` due to the [recent changes regarding `fuse_loop_config`](https://github.com/libfuse/libfuse/releases/tag/fuse-3.12.0).

Now I'm wondering what is the intended strategy to deal with incompatible API changes? Should we branch depending on `fuse_version()`? Or should we ship our own libfuse3.so (hoping it will be compatible with the kernel) with the application?

**Comment by bsbernd** [comment]
What exactly fails? Unless you are using ulibc or macos (apple) it is supposed to be compatible (versioned symbols).

**Reply by bsbernd** [reply]
Ah maybe and older compiler and you need this commit? https://github.com/libfuse/libfuse/commit/b1290d4c091a5c590e61a6195e625eab33300673

**Reply by overheadhunter** [reply]
Honestly, I have no idea how symbol lookups work in my case. The application is not statically linked and as far as I can tell, the linker looks for a symbol named `fuse_loop_mt` with no version suffixes.

I am using some experimental technology, though: Using jextract to generate Java interfaces from C header files. [This is the failing code](https://github.com/cryptomator/jfuse/blob/0.2.0/jfuse-linux-amd64/src/main/java/org/cryptomator/jfuse/linux/amd64/FuseMountImpl.java#L16-L19) that used to work fine with older libfuse versions and [here](https://github.com/cryptomator/jfuse/blob/0.2.0/jfuse-linux-amd64/src/main/java/org/cryptomator/jfuse/linux/amd64/extr/constants%241.java#L37-L40) and [here](https://github.com/cryptomator/jfuse/blob/0.2.0/jfuse-linux-amd64/src/main/java/org/cryptomator/jfuse/linux/amd64/extr/RuntimeHelper.java#L58-L62) are the relevant piece of machine-generated code to look up the symbol at runtime.

**Reply by overheadhunter** [reply]
Don't get me wrong, I can easily change the code to work with 3.12, but then I would rely on the new `fuse_loop_cfg_*` methods that won't work when the user has an older libfuse .so.

**Reply by bsbernd** [reply]
Which linux distribution and compile are you on? If you are familiar with C, can you please test if your compiler has "__attribute__ ((symver (sym2)?  Or just check how libfuse gets compiled, there is a meson test for it, see meson.build
```# gcc-10 and newer support the symver attribute which we need to use if we
# want to support LTO
# recent clang and gcc both support __has_attribute (and if they are too old
# to have __has_attribute, then they are too old to support symver)
# other compilers might not have __has_attribute, but in those cases
# it is safe for this check to fail and for us to fallback to the old _asm_
# method for symver. Anyway the attributes not supported by __has_attribute()
# unfortunately return true giving a false positive. So let's try to build
# using __attribute__ ((symver )) and see the result.
code = '''
__attribute__ ((symver ("test@TEST")))
void foo(void) {
}

int main(void) {
    return 0;
}'''
if cc.compiles(code, args: [ '-O0', '-c', '-Werror'])
     message('Compiler supports symver attribute')
     add_project_arguments('-DHAVE_SYMVER_ATTRIBUTE', language: 'c')
else
     message('Compiler does not support symver attribute')
endif
```

**Reply by bsbernd** [reply]
Or you could also just test laster libfuse master, which has the typo fix patch for compiler that do _not_ have `__attribute__ ((symver))`

**Reply by overheadhunter** [reply]
I have to ask about the distro, will report back.

However I have to stress (maybe I wasn't clear about it): At no time is there any static compilation. The header files are only used to derive a representation of the memory layouts and symbol names. I need to ask on the Java mailing list, how the "SymbolLookup" behaves regarding versioned symbols.

**Reply by overheadhunter** [reply]
@aakefbs I got replies on the mailing list:

> [...] symbol versioning is a GNU thing which may not exist in other implementations [...]

and

> Sadly, `dlopen` only does unversioned lookups [...]

So no symver support. Which brings me back to my question: What is the intended way to deal with API-breaking changes? Do we need to branch the code depending on `fuse_version()`?

**Reply by bsbernd** [reply]
Hmm ok, they use dlopen. I was just searching myself and found this reference https://stackoverflow.com/questions/5015367/how-to-achieve-symbol-versioning

Given that is a gnu extension - I do not know how to handle in in java. I guess one would have to use ifdef in the related code.

**Reply by bsbernd** [reply]
I really do not know how to automatically handle it for you - you would basically need to install multiple versions of libfuse and then use the 'right' one. I guess that means increasing the so.x version - which is exactly what multiple symbol versions is trying to avoid.

Manual handling is to call `fuse_session_loop_mt_32` and `fuse_parse_cmdline_30`.

**Reply by overheadhunter** [reply]
We just did some promising experiments with invoking `dlvsym` manually, however it is debatable, if the added complexity is worth it, when the alternative looks a rather simple (at least for this specific case):

```
if (fuse_version() < 312) {
  fuse_session_loop_mt(...)
} else {
  fuse_session_loop_mt_32(...)
}
```

The downside of such conditional invocations is that we need to update the project asap after a new libfuse version is released. While with `dlvsym` we can rely on the "old" function being called.

**Comment by bsbernd** [comment]
Isn't the easiest just to compile libfuse master, install it to <path/somewhere> and then let your java application test this <path/somewhere/libfuse.so>?

**Reply by overheadhunter** [reply]
I don't understand. Do you expect master to contain a fix, that isn't present in 3.12.0?

**Comment by bsbernd** [comment]
Yes, the issue was only reported after 3.12.0 was released already and so it got only fixed after the release.

**Reply by overheadhunter** [reply]
we can give it a try. However we did not encounter any problems with `fuse_parse_cmdline`, if you're referring to https://github.com/libfuse/libfuse/commit/b1290d4c091a5c590e61a6195e625eab33300673. The only problems arise when passing the "old" manually created `fuse_loop_config` to `fuse_loop_mt`. Using NULL instead works as intended.

**Comment by bsbernd** [comment]
Ah hmm, too many versioned symbols, I mixed them up. 
These version issues is exactly why I was not too happy with the previous api - every new option required a new versioned symbol.

Well I don't know, if jextract extracts the wrong version - no idea what to do - fix jextract?

**Reply by overheadhunter** [reply]
I will ask on the Java mailing list, if and how it supports versioned symbols

**Reply by overheadhunter** [reply]
Just so I get it right: Using libfuse 3.12, I can still pass an "old" `fuse_loop_config` (i.e. allocated manually without `fuse_loop_cfg_create`) to `fuse_loop_mt`, _if and only if_ linked against the correct version of `fuse_loop_mt`. Right?

**Comment by bsbernd** [comment]
See the example file systems, they define FUSE_USE_VERSION - below FUSE_MAKE_VERSION(3, 12) it uses the old API, see fuse_lowlevel.h

```
#if FUSE_USE_VERSION < 32
	int fuse_session_loop_mt_31(struct fuse_session *se, int clone_fd);
	#define fuse_session_loop_mt(se, clone_fd) fuse_session_loop_mt_31(se, clone_fd)
#elif FUSE_USE_VERSION < FUSE_MAKE_VERSION(3, 12)
	int fuse_session_loop_mt_32(struct fuse_session *se, struct fuse_loop_config *config);
	#define fuse_session_loop_mt(se, config) fuse_session_loop_mt_32(se, config)
#else
	#if (!defined(__UCLIBC__) && !defined(__APPLE__))
		/**
		 * Enter a multi-threaded event loop.
		 *
		 * For a description of the return value and the conditions when the
		 * event loop exits, refer to the documentation of
		 * fuse_session_loop().
		 *
		 * @param se the session
		 * @param config session loop configuration
		 * @return see fuse_session_loop()
		 */
		int fuse_session_loop_mt(struct fuse_session *se, struct fuse_loop_config *config);
	#else
		int fuse_session_loop_mt_312(struct fuse_session *se, struct fuse_loop_config *config);
		#define fuse_session_loop_mt(se, config) fuse_session_loop_mt_312(se, config)
	#endif
#endif
```

When compiled against an older libfuse and now the new libfuse is installed, it will use the compat  fuse_session_loop_mt symbol (fuse_session_loop_mt@FUSE_3.2  - ABI compatibility).

**Reply by overheadhunter** [reply]
That requires the versioning thing to work. Otherwise I could of course tell my application to explicitly use `fuse_session_loop_mt_32`, but I assume that this symbol is unknown in earlier versions.

Btw shouldn't it be `int fuse_session_loop_mt_32(struct fuse_session *se, struct fuse_loop_config_v1 *config);` (note the `_v1`)?

**Comment by bsbernd** [comment]
See fuse_common.h
```
/**
 * Configuration parameters passed to fuse_session_loop_mt() and
 * fuse_loop_mt().
 * Deprecated and replaced by a newer private struct in FUSE API
 * version 312 (FUSE_MAKE_VERSION(3, 12)
 */
#if FUSE_USE_VERSION < FUSE_MAKE_VERSION(3, 12)
struct fuse_loop_config_v1; /* forward declarition */
struct fuse_loop_config {
#else
struct fuse_loop_config_v1 {
#endif
```

So `struct fuse_loop_config` should work.
